### PR TITLE
feat: migrate dashboards from file provisioning to Grafana HTTP API

### LINKED ISSUE #170: Migrate dashboards from file provisioning to Grafana HTTP API
## What's wrong

Dashboards are file-provisioned (mounted read-only into the Grafana container). This means:

- Provisioned dashboards **cannot be updated via the Grafana API** — any change requires editing the JSON file on disk and restarting Grafana
- Agents editing dashboards must work blind with raw JSON and have no way to validate changes without a manual deploy
- The provisioning system has caused multiple outages: UID mismatches between provisioning YAML and Grafana's DB, restart-order sensitivities, and DNS failures from `docker stop/start` vs `down/up` cycles
- Every dashboard change requires a full container restart, which briefly takes down Grafana

Switching to API-managed dashboards keeps JSON files in git as the source of truth, but delivers them via HTTP API instead of file mount. No restart required for dashboard changes. Agents can validate by running the sync script.

## What done looks like

1. **`scripts/sync-dashboards.sh`** — idempotent script that pushes all JSON files from `stacks/observability/dashboards/` to Grafana via the HTTP API. Handles: Grafana readiness wait, folder creation, dashboard upsert with `overwrite: true`. Uses basic auth with `GRAFANA_ADMIN_PASSWORD` env var.

2. **`scripts/deploy.sh`** — add an `observability` case: run `docker compose up -d --remove-orphans`, then call `scripts/sync-dashboards.sh` to push dashboards after Grafana is healthy.

3. **`stacks/observability/compose.yaml`** — remove the `./dashboards:/var/lib/grafana/dashboards:ro` volume mount. Keep `./provisioning:/etc/grafana/provisioning:ro` (datasources and alerting stay file-provisioned).

4. **`stacks/observability/provisioning/dashboards/`** — delete the entire directory (the `dashboards.yaml` provider config is no longer needed).

5. **`.github/instructions/dashboards.instructions.md`** — instruction file scoped to `stacks/observability/dashboards/**` documenting: how dashboards are API-managed, the sync script, datasource UIDs (`prometheus`, `loki`), key query patterns (dedup with `max by (name)` / `max()`, job labels `integrations/unix` and `docker`), container name regex `(.+)-1$`, and how to test changes.

6. **`mise.toml`** — add `sync:dashboards` task that runs `scripts/sync-dashboards.sh`.

7. **Dashboard JSON cleanup** — each dashboard JSON file must have `"id": null` so Grafana matches by UID on upsert. Verify all 3 dashboards (`container-overview.json`, `agent-tasks.json`, `security.json`).

8. **`docs/observability.md`** — updated to reflect dashboards are now API-managed, with the sync command.

9. All CI checks pass (`mise run ci`).

### Grafana Dashboard API

Push a dashboard (create or update by UID):

```
POST http://100.100.146.119:3001/api/dashboards/db
Content-Type: application/json
Authorization: Basic <base64(admin:password)>

{
  "dashboard": { <full dashboard JSON with "id": null> },
  "folderUid": "homelab",
  "overwrite": true,
  "message": "Synced from git"
}
```

Response: `200 OK` with `{"id": N, "uid": "...", "url": "...", "status": "success"}`.

The `"id": null` is critical — if a dashboard JSON has a numeric ID from a previous provisioned state, the API will try to match by ID instead of UID, causing failures or updating the wrong dashboard.

### Grafana Folder API

Create the target folder (returns 409 if it exists — that's fine, ignore it):

```
POST http://100.100.146.119:3001/api/folders
Content-Type: application/json

{ "uid": "homelab", "title": "Homelab" }
```

### Grafana Health Endpoint

```
GET http://100.100.146.119:3001/api/health
```

Returns `200` when ready. Use this to wait for Grafana after compose up.

### Auth

Basic auth with curl: `curl -u "admin:${GRAFANA_ADMIN_PASSWORD}"`. The password is already available as `GRAFANA_ADMIN_PASSWORD` in the deploy environment (sourced from `.env`).

### Network

Grafana is bound to **`100.100.146.119:3001`** (Tailscale interface only). `localhost:3001` will NOT work — the port is not bound to `127.0.0.1`. All scripts must use `http://100.100.146.119:3001`.

### Migration order at deploy time

1. `docker compose up -d` starts Grafana **without** the dashboard volume mount and without the provisioning config → provisioned dashboards disappear
2. Sync script waits for Grafana health check
3. Sync script pushes dashboards via API → dashboards reappear as API-managed

There is a brief window (seconds) where dashboards are unavailable between steps 1 and 3. Acceptable for a homelab.

### Sync script pattern

```bash
#!/usr/bin/env bash
set -euo pipefail

GRAFANA_URL="http://100.100.146.119:3001"
GRAFANA_AUTH="admin:${GRAFANA_ADMIN_PASSWORD:?Set GRAFANA_ADMIN_PASSWORD}"
DASHBOARD_DIR="stacks/observability/dashboards"

# Wait for Grafana to be ready (up to 30s)
for i in $(seq 1 30); do
  if curl -sf "$GRAFANA_URL/api/health" > /dev/null 2>&1; then break; fi
  sleep 1
done

# Create folder (409 = already exists, that's fine)
curl -sf -u "$GRAFANA_AUTH" -X POST "$GRAFANA_URL/api/folders" \
  -H "Content-Type: application/json" \
  -d '{"uid": "homelab", "title": "Homelab"}' > /dev/null 2>&1 || true

# Push each dashboard
for f in "$DASHBOARD_DIR"/*.json; do
  name=$(basename "$f" .json)
  payload=$(jq -n --argjson dash "$(cat "$f")" \
    '{dashboard: ($dash | .id = null), folderUid: "homelab", overwrite: true, message: "Synced from git"}')
  curl -sf -u "$GRAFANA_AUTH" -X POST "$GRAFANA_URL/api/dashboards/db" \
    -H "Content-Type: application/json" \
    -d "$payload" > /dev/null
  echo "  📊 Synced $name"
done
```

This is a reference — the agent should implement and test this, handling edge cases (jq availability, error reporting, etc.).

## What must not break

- **Datasource provisioning** — file-based via `provisioning/datasources/datasources.yaml`, must remain unchanged
- **Alert rule provisioning** — file-based via `provisioning/alerting/`, must remain unchanged
- **All 3 dashboards** must be accessible after migration with the same UIDs and content
- **Deploy workflow** — `deploy.yaml` calls `deploy.sh`, which handles details. No workflow changes needed.
- **`mise run ci`** must pass
- **Dashboard datasource UIDs** — `prometheus` and `loki` must remain valid references

## Out of scope (no changes needed)

- `.github/workflows/` — no workflow changes required, `deploy.yaml` already calls `deploy.sh`
- Agent Python service — this is a deployment/tooling concern, not an agent app change

### PR DESCRIPTION
**Pull request by ColinCee**
## Summary

Replaces file-based dashboard provisioning with API-managed dashboards. JSON files stay in git as source of truth, pushed to Grafana via HTTP API during deploy.

Closes #170

## Changes

- **`scripts/sync-dashboards.sh`** — idempotent sync: waits for Grafana health, creates folder, upserts each dashboard by UID
- **`scripts/deploy.sh`** — observability case runs sync after compose up
- **`stacks/observability/compose.yaml`** — removed `./dashboards:/var/lib/grafana/dashboards:ro` mount
- **`stacks/observability/provisioning/dashboards/`** — deleted (provider config no longer needed)
- **`mise.toml`** — added `sync:dashboards` task
- **`.github/instructions/dashboards.instructions.md`** — agent instruction file with query patterns, dedup rules, datasource UIDs
- **`docs/observability.md`** — updated to reflect API-managed dashboards

## Verified

Tested full migration on server:
1. Removed provisioning config + volume mount → restarted Grafana
2. Ran sync script → all 3 dashboards pushed successfully
3. Verified idempotency (re-ran sync, no errors)
4. All dashboards accessible in Grafana

### CODE REVIEW

File: scripts/sync-dashboards.sh [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+# Idempotent — safe to run repeatedly. Dashboards are matched by UID.
+#
+# Requires: curl, python3 (for JSON envelope construction)
+# Environment: GRAFANA_ADMIN_PASSWORD (required)
+#
+# Usage: sync-dashboards.sh
+#    or: GRAFANA_URL=http://host:3001 sync-dashboards.sh
+
+GRAFANA_URL="${GRAFANA_URL:-http://100.100.146.119:3001}"
+GRAFANA_AUTH="admin:${GRAFANA_ADMIN_PASSWORD:?Set GRAFANA_ADMIN_PASSWORD}"
```
[the remark this answers is not in the document]
Comment (ColinCee) [reply scripts/sync-dashboards.sh:14]: Fixed in 6ad328f: `GRAFANA_ADMIN_PASSWORD` added to `.env.example` so `generate-env.sh` wires it through. GHA secret created. The deploy workflow exports all secrets before `deploy.sh` runs.

File: mise.toml [hunk 2]
```diff
@@ -90,6 +90,11 @@ run = "docker compose -f stacks/agents/compose.yaml up -d --build"
 description = "Deploy flight tracker (pull from GHCR)"
 run = "docker compose -f stacks/flight-tracker/compose.yaml pull && docker compose -f stacks/flight-tracker/compose.yaml up -d"
 
+# ── Sync ───────────────────────────────────────────
+[tasks."sync:dashboards"]
+description = "Push dashboard JSON files to Grafana via API"
+run = "scripts/sync-dashboards.sh"
```
[the remark this answers is not in the document]
Comment (ColinCee) [reply mise.toml:96]: Fixed in 6ad328f: `deploy:observability` mise task now runs `sync-dashboards.sh` after compose up, same as `deploy.sh`.
